proxy/http2/Http2ConnectionState.cc

```diff
@@ -1189,6 +1217,93 @@ Http2ConnectionState::send_headers_frame(Http2Stream *stream)
 }
 
 void
+Http2ConnectionState::send_push_promise_frame(Http2Stream *stream, const char *url, int url_len)
+{
+  HTTPHdr h1_hdr, h2_hdr;
+  uint8_t *buf                = NULL;
+  uint32_t buf_len            = 0;
+  uint32_t header_blocks_size = 0;
+  int payload_length          = 0;
+  uint64_t sent               = 0;
+  uint8_t flags               = 0x00;
+
+  if (client_settings.get(HTTP2_SETTINGS_ENABLE_PUSH) == 0) {
+    return;
+  }
+
+  DebugHttp2Stream(ua_session, stream->get_id(), "Send PUSH_PROMISE frame");
+
+  h1_hdr.create(HTTP_TYPE_REQUEST);
+  h1_hdr.url_set(url, url_len);
+  h1_hdr.method_set("GET", 3);
+  http2_generate_h2_header_from_1_1(&h1_hdr, &h2_hdr);
+
+  buf_len = h1_hdr.length_get() * 2; // Make it double just in case
+  h1_hdr.destroy();
+  buf = (uint8_t *)ats_malloc(buf_len);
+  if (buf == NULL) {
+    h2_hdr.destroy();
+    return;
+  }
+  Http2ErrorCode result = http2_encode_header_blocks(&h2_hdr, buf, buf_len, &header_blocks_size, *(this->remote_hpack_handle));
+  if (result != HTTP2_ERROR_NO_ERROR) {
+    h2_hdr.destroy();
+    ats_free(buf);
+    return;
+  }
+
+  // Send a PUSH_PROMISE frame
+  if (header_blocks_size <= BUFFER_SIZE_FOR_INDEX(buffer_size_index[HTTP2_FRAME_TYPE_PUSH_PROMISE]) - HTTP2_FRAME_HEADER_LEN) {
```

Comment:
`HTTP2_FRAME_TYPE_PUSH_PROMISE` is used for frame payload size in below (L1263), so it doesn't include frame header size.
If I understand correctly, we should minus 4 byte (length of Promised Stream ID).